producer.py

```diff
@@ -0,0 +1,166 @@
+                index_url,
+            )
+            packages.append((package_name, version, index_url, solver_name))
+
+    if not packages:
+        _LOGGER.info("No unsolved packages found")
+        return
+
+    # Shuffle not to be dependent on solver message ordering.
+    random.shuffle(packages)
+
+    revsolver_packages_seen = set()
+    async_tasks = []
+    # Class for solver messages
+    unresolved_package = UnresolvedPackageMessage()
+    # Class for reverse solver messages
+    unrevsolved_package = UnrevsolvedPackageMessage()
+
+    for package_name, package_version, index_url, solver_name in packages:
+        for index_url in [index_url] if index_url is not None else indexes:
+            try:
+                async_tasks.append(
+                    unresolved_package.publish_to_topic(
+                        unresolved_package.MessageContents(
+                            package_name=package_name,
+                            package_version=package_version,
+                            index_url=[index_url],
+                            solver=solver_name,
+                        )
+                    )
+                )
+                _LOGGER.info(
+                    "Scheduled solver %r for package %r in version %r from index %r, analysis is %r",
+                    solver_name,
+                    package_name,
+                    package_version,
+                    index_url,
+                )
+            except Exception as identifier:
+                _LOGGER.exception(
+                    "Failed to publish solver message with the following error message: %r",
+                    identifier,
+                )
+
+            # Send reverse solver message if not done for this packge, package_version
+            if (package_name, package_version) not in revsolver_packages_seen:
+                try:
+                    async_tasks.append(
+                        unrevsolved_package.publish_to_topic(
+                            unrevsolved_package.MessageContents(
+                                package_name=package_name,
+                                package_version=package_version,
+                            )
+                        )
+                    )
+                    _LOGGER.info(
+                        "Scheduled reverse solver message for package %r in version %r, analysis is %r",
+                        package_name,
+                        package_version,
+                    )
+                    revsolver_packages_seen.add((package_name, package_version))
+                except Exception as identifier:
+                    _LOGGER.exception(
+                        "Failed to publish reverse solver message with the following error message: %r",
+                        identifier,
+                    )
+
+    # Finally gather all the async co-routines
+    await asyncio.gather(*async_tasks)
+
+
+if __name__ == "__main__":
+    asyncio.run(main())
```

Comment:
```
if __name__ == "__main__":
    loop = asyncio.get_event_loop()
    loop.run_until_complete(main())
```

Reply:
This seems the recommended way - https://docs.python.org/3.8/whatsnew/3.8.html#asyncio

Reply:
Any reason to move to Python 3.8? The whole application runs on Python 3.6. (we will schedule migration to 3.8 at some point though.)

Reply:
I thought we we were restructuring, tried upgrading to Python 3.8, though I can change it back to Python 3.6, should we do that?

Reply:
> I thought we we were restructuring, tried upgrading to Python 3.8, though I can change it back to Python 3.6, should we do that?

I think having all Python3.6 for now makes sense. We can switch to Python 3.8 later on, we haven't tested the whole application (and all the transitive deps). Let's do it in a separate step.

Reply:
That makes sense, thanks. I will change it :)